### PR TITLE
Archiving task from experiment table - connections are lost [SCI-7732]

### PR DESCRIPTION
**Pull request by sboursen-scinote**
Jira ticket: [SCI-7732](https://scinote.atlassian.net/browse/SCI-7732)

### What was done
I prevented connection loss after an intermediary `task` (`my_module`) is archived. Now the connection is maintained between the `src` tasks and the `dest` tasks following the same logic in `Canvas` (two-by-two connections between src tasks and dest tasks)

[SCI-7732]: https://scinote.atlassian.net/browse/SCI-7732?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/controllers/experiments_controller.rb [hunk 1]
```diff
@@ -630,4 +633,22 @@ def sort_my_modules(records, sort)
       records
     end
   end
+
+  def generate_two_by_two_connections(src_tasks, dest_tasks)
+    return if src_tasks.empty? || dest_tasks.empty?
+
+    src_tasks.each do |src|
```
Comment (artoscinote) [inline app/controllers/experiments_controller.rb:640]: The MyModule model already has these relation methods defined:
```
  has_many :my_modules, through: :outputs, source: :to, class_name: 'MyModule'
  has_many :my_module_antecessors, through: :inputs, source: :from, class_name: 'MyModule'
```

so we can just use these here

File: app/controllers/experiments_controller.rb [hunk 2]
```diff
@@ -517,7 +520,7 @@ def view_type_params
   def check_read_permissions
     current_team_switch(@experiment.project.team) if current_team != @experiment.project.team
     render_403 unless can_read_experiment?(@experiment) ||
-                      @experiment.archived? && can_read_archived_experiment?(@experiment)
+                      (@experiment.archived? && can_read_archived_experiment?(@experiment))
```
Comment (artoscinote) [inline app/controllers/experiments_controller.rb:523]: was this change intentional?

File: app/controllers/experiments_controller.rb [hunk 3]
```diff
@@ -630,4 +633,22 @@ def sort_my_modules(records, sort)
       records
     end
   end
+
+  def generate_two_by_two_connections(src_tasks, dest_tasks)
```
Comment (artoscinote) [inline app/controllers/experiments_controller.rb:637]: should maybe be more descriptive, maybe something like `reconnect_my_modules_after_archive`

File: app/controllers/experiments_controller.rb [hunk 4]
```diff
@@ -630,4 +633,22 @@ def sort_my_modules(records, sort)
...
+    return if src_tasks.empty? || dest_tasks.empty?
+
+    src_tasks.each do |src|
+      dest_tasks.each do |dest|
+        Connection.create!(input_id: dest.id, output_id: src.id)
+      end
+    end
+  end
+
+  def get_dest_my_modules(my_module)
```
Comment (artoscinote) [inline app/controllers/experiments_controller.rb:647]: These methods are not needed (as we already have these defined on MyModule), but as a general rule, in Ruby we don't use `get_` prefixes, as they are redundant. In this case, a more proper name would just be `destination_my_modules`.

**Review by artoscinote — CHANGES_REQUESTED** [review]

File: app/controllers/experiments_controller.rb (hunk 1 above)
Comment (sboursen-scinote) [reply app/controllers/experiments_controller.rb:640]: Hi Martin,

Thank you for mentioning that.
Will work on it right away.

File: app/controllers/experiments_controller.rb (hunk 2 above)
Comment (sboursen-scinote) [reply app/controllers/experiments_controller.rb:523]: No, I think rubocop recommended it!
Should I change it back to how it was?

**Review by artoscinote — APPROVED** [review]
